### PR TITLE
Performance improvement to improve training time

### PR DESCRIPTION
**Pull request by edkazcarlson**
In this PR I am cleaning up some existing code to help improve training times.
The main speedups come from not running transforms that are passed with a 0% chance of applying their transform but I also introduced jit compiled methods with the numba library to help handle some operations that are ran frequently.
If the team doesn't want to use numba (code clutter, licensing, etc) I'll remove as while they did introduce some speedup, a majority of the speedup was from the changes to the image transformations.

Tested with the following code
```python
start_time = time.time()

modelName = f'yolov8n.yaml' 
overrides = {'epochs': 6, 'imgsz': 640, 'data': 'coco.yaml', 'model': modelName, 'batch': 8, 'close_mosaic': 3}
trainer = DetectionTrainer(overrides=overrides)
trainer.train()

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} seconds")
```

image
Overall with my changes it took 9754 seconds to train for 6 epochs (3 with mosaic, 3 without) where it would take 10531 without my changes.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced and streamlined data augmentation and instance handling in the Ultralytics repository.

### 📊 Key Changes
- **Transformation Updates:** Refactored the `v8_transforms` function to conditionally add transformations based on hyperparameters.
- **Instance Scaling:** Introduced `scaleWH` and `scaleSingle` methods for more precise scaling of bounding boxes and other components.
- **Batch Collation:** Improved the `collate_fn` for batches by optimizing how batch indices are handled.
- **Loss Improvements:** Optimized the loss calculations for better performance and consistency.
- **Operations Enhancement:** Simplified and made conversion functions (`xyxyxyxy2xywhr` and `scale_coords`) more efficient.

### 🎯 Purpose & Impact
- **Enhanced Flexibility:** Users can now enable/disable specific transformations based on configuration, providing more control over data augmentation. 🌱
- **Improved Accuracy:** New scaling methods ensure precise adjustments of bounding boxes and other instances, reducing potential errors in processing. 🎯
- **Performance Boost:** Optimized batch collation and loss computation, leading to potentially faster training times and smoother performance. 🚀
- **Simplified Code:** Cleaner and more maintainable codebase, making it easier for developers to understand and contribute. 📚

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Thanks for this pull request and your efforts to enhance the training performance of the YOLOv8 models! 🚀 The optimizations to the image transformations seem particularly promising as they effectively cut down operational redundancy. Using numba for jit compilation is an intriguing choice, and while it does add external dependency, the speed improvements you've observed might justify its use.

Could you provide some benchmarks comparing the performance with and without the use of numba? This data could help in making a more informed decision about integrating numba more broadly within the codebase.

Here’s a slight modification to the test snippet to include timing for both scenarios:

```python
import time
import numba  # Remove if not using numba

def train_model(use_numba):
    start_time = time.time()

    modelName = 'yolov8n.yaml'
    overrides = {'epochs': 6, 'imgsz': 640, 'data': 'coco.yaml', 'model': modelName, 'batch': 8, 'close_mosaic': 3}
    trainer = DetectionTrainer(overrides=overrides)
    trainer.train()

    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Execution time with{'out' if not use_numba else ''} numba: {execution_time} seconds")

train_model(use_numba=True)
train_model(use_numba=False)
```

This modification helps to directly compare the impact of numba on the overall training process. Looking forward to your findings!

**Comment by glenn-jocher** [comment]
@Laughing-q interesting training speedup PR. I think we'd like to merge this without the `numba` addition as it may be hardware specific and we'd strongly prefer to avoid adding additional dependencies.

**Comment by edkazcarlson** [comment]
image
slightly slower than before when removing numba but still faster than the current state of main

**Comment by glenn-jocher** [comment]
Thanks for the update! It's great to hear that the performance improved even without numba. If you can share the specific metrics or any additional insights from your latest tests, that would be helpful for finalizing the merge. Let's aim for the best balance of dependency minimization and performance enhancement. 🚀

**Comment by edkazcarlson** [comment]
> Thanks for the update! It's great to hear that the performance improved even without numba. If you can share the specific metrics or any additional insights from your latest tests, that would be helpful for finalizing the merge. Let's aim for the best balance of dependency minimization and performance enhancement. 🚀

In short my main changes currently are:
1. Taking better advantage of vectorized methods in order to boost perf 
2. Not applying transforms if they are passed a 0% probability of running, thus saving us (# of transforms x images x epochs x cpu clocks it takes to run random.random())

I don't have any specific metrics outside of wallclock time for the epochs. Is there something specific the team would want?
Thanks :)

**Comment by glenn-jocher** [comment]
Thanks for detailing the changes! The approach sounds solid, particularly your method to skip transformations when their probability is zero—definitely a smart optimization. 😊

For metrics, if you could provide us with a comparison in wall-clock time between the main branch and your changes (i.e., how much total time each epoch takes on average) across a few runs, that would be ideal. This will help us quantitatively assess the impact of your improvements.

Keep up the fantastic work! Looking forward to integrating these enhancements. 🌟

**Comment by Laughing-q** [comment]
@edkazcarlson Hi, thanks for the PR! 
I tested your changed locally but the results are almost the same as main branch. Here's my results:
on main branch:
pic-240513-1946-15
on current PR:
pic-240513-2037-44
And my testing command:
```bash
yolo train detect data=coco.yaml model=yolov8m.yaml batch=64 epochs=4 close_mosaic=2 device=0,1,2,3
```

**Comment by edkazcarlson** [comment]
@Laughing-q  Thank you for the tests on your hardware, could you try doing this through python itself? Not sure where exactly the yolo command is pointing, could it be that it's pointing at the install you have through pip and not my branch (can you confirm through using `which` ?)
I haven't had much of a chance between work and other stuff so haven't gotten much of a chance to do in depth tests but from what I can see just initially comparing my old tests to some new perf tests it seems like the new main is slower, checking out an old commit (ex: 1365fe9d6cf5ad70a57c99bc1c193d0664dae896 ) seems to be faster than the branch thats merged w/ main.
I plan to keep investigating this slowdown in the merged commits but should we merge this into main for the time being? I don't see this PR hurting perf in any way (can change title just for better record keeping)

**Comment by edkazcarlson** [comment]
Also side question, I know numba got denied due to hardware compat reasons, but does the team accept cython improvements?

**Comment by glenn-jocher** [comment]
Hi there! Thanks for your continued contributions and for checking in about Cython. Yes, we're open to considering Cython improvements as they can be a great way to enhance performance while maintaining compatibility across various hardware setups. If you have specific optimizations in mind using Cython, feel free to share them or open a PR. We'd love to take a look! 🚀

**Comment by edkazcarlson** [comment]
Thank you for your patience, after a number of different tests, I realized that a huge portion of my performance gains were likely caused by the (slightly) better cooling that my laptops stand provides, where I was getting 6-8% better performance w/ the laptop stand than w/o. Slightly embarrassing but I'll try to keep this in mind for any future performance work I do, apologies for the initial confusion and overblown estimate.

When comparing w/ my changes and w/o my changes though, I am still getting ~5% faster with laptop stand and ~3.5% without laptop stand.

I used the following code (6 epochs, 3 with mosaic, 3 without) in order to test.
With my changes, my times across multiple sessions were 
(without laptop stand)
10131 seconds
9682 
10084
for an average of 9965 seconds
while the main branch got
10390
10263
for an average of 10326sec, meaning my changes took (9965 / 10326) = 96.503970559% of the duration main did

For completeness, with the stand my changes took 9223s and 9311 (avg 9267) vs main having 9697s and 9804s (avg 9750s), meaning with my changes it took 95.046153846% as long
```
from ultralytics.models.yolo.detect import DetectionTrainer
import time

start_time = time.time()

modelName = f'yolov8n.yaml' 
overrides = {'epochs': 6, 'imgsz': 640, 'data': 'coco.yaml', 'model': modelName, 'batch': 8, 'close_mosaic': 3}
trainer = DetectionTrainer(overrides=overrides)
trainer.train()

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} seconds")
```

**Comment by glenn-jocher** [comment]
Thank you for the detailed update and for your honesty about the cooling factor—it's an interesting observation that highlights how various environmental factors can impact performance testing. 🌡️

Your continued efforts and the results you've shared are valuable. A consistent 3.5% to 5% improvement in training time is still quite significant, especially when scaled across multiple training sessions and models. It's clear that your changes are having a positive impact, even if the initial estimates were influenced by external factors.

Let's proceed with integrating your changes into the main branch. This will allow us to benefit from these improvements and also keep the project moving forward efficiently. Great work, and looking forward to more of your contributions! 🚀

**Comment by edkazcarlson** [comment]
@glenn-jocher @Laughing-q Should I just consider this PR a wash? I can keep re-merging to infinity (and I'll probably set up a cron job soon) but I assume the PR validation isn't free for you.

**Comment by edkazcarlson** [comment]
Same w/ this https://github.com/ultralytics/ultralytics/pull/10308 PR

**Comment by glenn-jocher** [comment]
Thank you for your patience. We'll review the PR and provide feedback soon.
